core/domain/exp_domain.py

```diff
@@ -474,6 +474,258 @@ def __init__(self, change_list):
+        for version, (name, _) in self:
+            yield (version, name)
+
+    def iterstates(self):
+        """Yields all (version, state) pairs that this span holds."""
+        for version, (_, state) in self:
+            yield (version, state)
+
+    def get_version(self, version):
+        """Returns the state as it appeared in the given exploration version.
+
+        Args:
+            version: int. The version to get.
+
+        Returns:
+            state_domain.State. The state as it appeared in the given version.
+        """
+        _, state = self[version]
+        return state
+
+    def get_version_name(self, version):
+        """Returns the name of the state at the given exploration version.
+
+        Args:
+            version: int. The version to get.
+
+        Returns:
+            str. The name of the state at the given exploration version.
+        """
+        name, _ = self[version]
+        return name
+
+    def get_multi_versions(self, version_start, version_end):
+        """Returns states for a half-open range of exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(state_domain.State). The states corresponding to the given
+            range of versions.
+        """
+        return [
+            state
+            for _, state in self._get_multi_versions(version_start, version_end)
+        ]
+
+    def get_multi_version_names(self, version_start, version_end):
+        """Returns state names for a half-open range of exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(str). The names corresponding to the given range of versions.
+        """
+        return [
+            name
+            for name, _ in self._get_multi_versions(version_start, version_end)
+        ]
+
+    def extend_or_split(
+            self, exp_version, state_name, state, exp_version_diff,
+            split_predicate=None):
+        """Extends span to include the given state, unless it fails to pass the
+        given split predicate, in which case a new span is created and returned.
+
+        Args:
+            exp_version: int. The state's corresponding exploration version.
+            state_name: str. The state's name at the given version.
+            state: state_domain.State. State's representation at given version.
+            exp_version_diff: ExplorationVersionsDiff. The changes since the
+                last exploration version.
+            split_predicate:
+                callable(state_domain.State, state_domain.State) -> bool | None.
+                Should return True when the two states should be split up. If
```

Comment:
Maybe explain more what you mean here by "splitting up a state" -- I don't think that's quite what you're doing, right? Each state is still an integral thing -- rather, you're doing something with the span that contains the state, I think.

Reply:
OK, I've also renamed it and have given it more context. How's this?